txstratum/healthcheck.py

```diff
@@ -0,0 +1,265 @@
+    """This is an interface to be used by other classes implementing health checks for components."""
+
+    @abstractmethod
+    async def get_health_check(self) -> ComponentHealthCheck:
+        """Return the health check status for the component."""
+        raise NotImplementedError()
+
+
+class HealthCheck:
+    """This is the main class that will use the other classes to check the health of the components.
+
+    It will aggregate the responses into a final object to be returned following our standards.
+    """
+
+    def __init__(self, manager: "TxMiningManager", backend: "HathorClient") -> None:
+        """Init the class with the components that will be checked."""
+        self.health_check_components: List[ComponentHealthCheckInterface] = [
+            MiningHealthCheck(manager),
+            FullnodeHealthCheck(backend),
+        ]
+
+    async def get_health_check(self) -> HealthCheckResult:
+        """Return the health check status for the tx-mining-service."""
+        components_health_checks = [
+            await c.get_health_check() for c in self.health_check_components
+        ]
+        components_status = {c.status for c in components_health_checks}
+
+        overall_status = HealthCheckStatus.PASS
+        if HealthCheckStatus.FAIL in components_status:
+            overall_status = HealthCheckStatus.FAIL
+        elif HealthCheckStatus.WARN in components_status:
+            overall_status = HealthCheckStatus.WARN
+
+        return HealthCheckResult(
+            status=overall_status,
+            description="health of tx-mining-service",
+            checks={c.component_name: [c] for c in components_health_checks},
+        )
+
+
+class FullnodeHealthCheck(ComponentHealthCheckInterface):
+    """This class will check the health of the fullnode by sending a request to its /v1a/health."""
+
+    def __init__(self, backend: "HathorClient") -> None:
+        """Init the class with the fullnode backend."""
+        self.backend = backend
+
+    async def get_health_check(self) -> ComponentHealthCheck:
+        """Return the fullnode health check status."""
+        health_check = ComponentHealthCheck(
+            component_name="fullnode",
+            component_type=ComponentType.FULLNODE,
+            # TODO: Ideally we should not use private fields. We'll fix this when fixing line 170
+            component_id=self.backend._base_url,
+            status=HealthCheckStatus.PASS,
+            time=datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ"),
+            output="fullnode is responding correctly",
+        )
+
+        try:
+            # TODO: We need to get the health information from the fullnode, but it's not implemented yet
+            await self.backend.version()
+
+            return health_check
+        except Exception:
+            health_check.status = HealthCheckStatus.FAIL
+            health_check.output = "couldn't connect to fullnode"
+
+            return health_check
```

Comment:
Also, I'm not sure how the `HathorClient` class works, does it perform any check on the returned value? I mean, I know that some full node APIs can respond with 200 OK even if it fails, and the contents could be something like `{"success": false}`.

Maybe we should check some attribute that we know is returned from `/v1a/version`, just to be sure. For example, checking that we receive a `version` attribute (only while `/v1a/health` is not implemented).

Reply:
You can see the `HathorClient` in https://github.com/HathorNetwork/python-hathorlib/blob/main/hathorlib/client.py

The class just assumes the "happy path" and tries to extract the `version` attribute from the response and parse it.

So it kind of already checks the attribute, and if it's not there some kind of Exception will be raised and we will catch it.

I was considering this to be good enough while we don't implement the `/v1a/health` on hathor-core, which should happen soon.